cleanlab/internal/neighbor/search.py

```diff
@@ -0,0 +1,37 @@
+from __future__ import annotations
+from typing import TYPE_CHECKING
+
+from sklearn.neighbors import NearestNeighbors
+
+
+if TYPE_CHECKING:
+
+    from cleanlab.internal.neighbor.types import Metric
+
+
+def construct_knn(n_neighbors: int, metric: Metric, **knn_kwargs) -> NearestNeighbors:
+    """
+    Constructs a k-nearest neighbors search object.
+
+    Parameters
+    ----------
+    n_neighbors :
+        The number of nearest neighbors to consider.
+    metric :
+        The distance metric to use for computing distances between points.
+        See :py:mod:`~cleanlab.internal.neighbor.metric` for more information.
+    **knn_kwargs:
+        Additional keyword arguments to be passed to the search index constructor.
+        See https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html for more details on the available options.
+
+    Returns:
+        NearestNeighbors: A k-nearest neighbors search object.
```

Comment:
where is the single source of truth across the entire library that dictates what attributes and what properties this NearestNeighbors object fundamentally must possess?

If there isn't a place yet, let's enumerate this here, and now before we forget to.  That knowledge would be crucial for changing the default KNN library in the future, for others to use their own KNN library, and for future devs adding issue-types that rely on KNN to understand what properties they are guaranteed to be able to rely on.

I assume the properties are simply a subset of the NearestNeighbors object that sklearn returns, much like CleanLearning can be run on any classifier that implements a subset of the properties of sklearn's base classifier object.

Reply:
I've now listed the necessary methods, the standard attributes as well as some some additional information about fitted attributes (those that are only set after calling `NearestNeighbors.fit()`)